megatron/model/init_functions.py

```diff
@@ -103,105 +84,91 @@ def init_(tensor):
+def xavier_normal_init_method(mup_m_width=1.0):
     """Fills the input Tensor with values according to the method described in Understanding the difficulty of
     training deep feedforward neural networks - Glorot, X. & Bengio, Y. (2010), using a normal distribution."""
 
-    def init_(tensor, use_mup=use_mup_outer):
-        if use_mup:
-            mup.init.xavier_normal_(tensor)
+    def init_(tensor, mup_m_width=mup_m_width):
+        init_weight = torch.nn.init.xavier_normal_(tensor)
+        if mup_m_width != 1:
             with torch.no_grad():
-                tensor.mul_(mup_init_scale)
-            return tensor
-        else:
-            return torch.nn.init.xavier_normal_(tensor)
+                init_weight.div_(mup_m_width)
+        return init_weight
 
     return init_
 
 
-def small_init_init_method(dim, use_mup_outer=False, mup_init_scale=1.0):
+def small_init_init_method(dim, mup_m_width=1.0):
     """Fills the input Tensor with values according to the method described in Transformers without Tears: Improving
     the Normalization of Self-Attention - Nguyen, T. & Salazar, J. (2010), using a normal distribution."""
     std = math.sqrt(2 / (5 * dim))
 
-    def init_(tensor, use_mup=use_mup_outer):
-        if use_mup:
-            mup.init.normal_(tensor, mean=0.0, std=std)
+    def init_(tensor, mup_m_width=mup_m_width):
+        init_weight = torch.nn.init.normal_(tensor, mean=0.0, std=std)
+        if mup_m_width != 1:
             with torch.no_grad():
-                tensor.mul_(mup_init_scale)
-            return tensor
-        else:
-            return torch.nn.init.normal_(tensor, mean=0.0, std=std)
+                init_weight.div_(mup_m_width)
+        return init_weight
 
     return init_
 
 
-def wang_init_method(n_layers, dim, use_mup_outer=False, mup_init_scale=1.0):
+def wang_init_method(n_layers, dim, mup_m_width=1.0):
     std = 2 / n_layers / math.sqrt(dim)
 
-    def init_(tensor, use_mup=use_mup_outer):
-        if use_mup:
-            mup.init.normal_(tensor, mean=0.0, std=std)
+    def init_(tensor, mup_m_width=mup_m_width):
+        init_weight = torch.nn.init.normal_(tensor, mean=0.0, std=std)
+        if mup_m_width != 1:
             with torch.no_grad():
-                tensor.mul_(mup_init_scale)
-            return tensor
-        else:
-            return torch.nn.init.normal_(tensor, mean=0.0, std=std)
-
+                init_weight.div_(mup_m_width)
+        return init_weight
+            
     return init_
 
 
 def get_init_methods(args):
 
-    if args.use_mup:
-        try:
-            import mup
-        except ModuleNotFoundError:
-            print("Please install mup https://github.com/microsoft/mup")
-            raise Exception
-
     def _get(name):
         if name == "normal":
             return init_method_normal(
-                args.init_method_std, args.use_mup, args.mup_init_scale
+                sigma=args.init_method_std/math.sqrt(args.mup_m_width)
```

Comment:
if you are going to define the muP init_std adjustment here then you need to remove all the `init_weight.div_` calls so you dont double adjust

Reply:
I'm split on what the best interface would be. For `init_method_normal` and `scaled_init_method_normal` I opted to not scale in the function but for other init functions I did.